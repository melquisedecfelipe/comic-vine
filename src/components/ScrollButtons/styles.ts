import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    position: fixed;
    bottom: 0;
    right: 0;
    display: flex;
    gap: ${theme.spacings.medium};
    padding: ${theme.spacings.large};
  `}
`
