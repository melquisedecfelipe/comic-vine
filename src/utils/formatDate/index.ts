export const formatDate = (date?: string | null): string => {
  if (!date) {
    return ''
  }

  const newDate = new Date(date)

  return new Intl.DateTimeFormat('en-US', {
    day: 'numeric',
    month: 'short',
    year: 'numeric'
  }).format(newDate)
}

export const formatDateToIso = (date?: string | null): string => {
  if (!date) {
    return ''
  }

  const newDate = new Date(date)

  return newDate.toISOString().substring(0, 10)
}
