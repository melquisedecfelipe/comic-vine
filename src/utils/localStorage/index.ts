import { Character, CharacterEdit } from 'types/characters'

export function getStorageItem(key: string) {
  if (typeof window === 'undefined') return

  const data = window.localStorage.getItem(key)

  return JSON.parse(data!)
}

export function setStorageItem(
  key: string,
  value: string[] | Character[] | CharacterEdit
) {
  if (typeof window === 'undefined') return

  const data = JSON.stringify(value)

  return window.localStorage.setItem(key, data)
}
