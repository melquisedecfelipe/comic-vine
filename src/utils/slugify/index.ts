export const slugify = (id: number, slug: string): string => {
  const string = slug.split(' ').join('-').toLocaleLowerCase()

  return [id, string].join('-')
}
