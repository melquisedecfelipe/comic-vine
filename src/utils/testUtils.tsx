import { ReactElement } from 'react'

import { render, RenderOptions } from '@testing-library/react'

import { ThemeProvider } from 'styled-components'

import {
  CharactersContext,
  CharactersContextData,
  CharactersContextDefaultValues
} from 'hooks/useCharacters'

import theme from 'styles/theme'

type CustomRenderProps = {
  charactersProviderProps?: CharactersContextData
} & Omit<RenderOptions, 'queries'>

const customRender = (
  ui: ReactElement,
  {
    charactersProviderProps = CharactersContextDefaultValues,
    ...renderOptions
  }: CustomRenderProps = {}
) =>
  render(
    <ThemeProvider theme={theme}>
      <CharactersContext.Provider value={charactersProviderProps}>
        {ui}
      </CharactersContext.Provider>
    </ThemeProvider>,
    renderOptions
  )

export * from '@testing-library/react'
export { customRender as render }
