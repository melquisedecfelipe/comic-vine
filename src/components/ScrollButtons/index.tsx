import Button from 'components/Button'

import * as S from './styles'

export type ScrollButtonsProps = {
  handleDown: () => void
  handleUp: () => void
}

const ScrollButtons = ({ handleDown, handleUp }: ScrollButtonsProps) => (
  <S.Wrapper>
    <Button aria-label="Down" variant="outlined" onClick={handleDown}>
      Down
    </Button>
    <Button aria-label="Up" variant="outlined" onClick={handleUp}>
      Up
    </Button>
  </S.Wrapper>
)

export default ScrollButtons
