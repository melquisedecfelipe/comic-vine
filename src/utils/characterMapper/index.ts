import { slugify } from 'utils/slugify'

import { BBFCharacter, Character, CharacterDetail } from 'types/characters'

const genders = new Map()

genders.set(0, 'Unknown')
genders.set(1, 'Male')
genders.set(2, 'Female')

export const characterMapper = (character: BBFCharacter): CharacterDetail => {
  return {
    aliases: character.aliases,
    detail: character.api_detail_url,
    birth: character.birth,
    gender: genders.get(character.gender),
    id: character.id,
    images: {
      icon: character.image.icon_url,
      screen: character.image.screen_url,
      small: character.image.small_url
    },
    name: character.name,
    realName: character.real_name,
    deck: character.deck,
    description: character.description,
    origin: character.origin.name,
    publisher: character.publisher.name,
    site: character.site_detail_url,
    slug: slugify(character.id, character.name)
  }
}

export const charactersMapper = (characters: BBFCharacter[]): Character[] => {
  return characters.map(character => ({
    aliases: character.aliases,
    detail: character.api_detail_url,
    birth: character.birth,
    id: character.id,
    gender: genders.get(character.gender),
    images: {
      icon: character.image.icon_url,
      screen: character.image.screen_url,
      small: character.image.small_url
    },
    name: character.name,
    realName: character.real_name,
    deck: character.deck,
    description: character.description,
    site: character.site_detail_url,
    slug: slugify(character.id, character.name)
  }))
}
