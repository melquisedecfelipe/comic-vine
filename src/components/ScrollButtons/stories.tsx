import { Story, Meta } from '@storybook/react/types-6-0'

import ScrollButtons, { ScrollButtonsProps } from '.'

export default {
  title: 'ScrollButtons',
  component: ScrollButtons,
  argTypes: {
    handleDown: { action: 'handleDownClick' },
    handleUp: { action: 'handleUpClick' }
  }
} as Meta

export const Default: Story<ScrollButtonsProps> = args => (
  <ScrollButtons {...args} />
)
