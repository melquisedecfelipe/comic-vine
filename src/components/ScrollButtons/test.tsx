import userEvent from '@testing-library/user-event'

import { render, screen } from 'utils/testUtils'

import ScrollButtons from '.'

const handleDown = jest.fn()
const handleUp = jest.fn()

const props = {
  handleDown,
  handleUp
}

describe('<ScrollButtons />', () => {
  it('should render the scroll buttons', () => {
    const { container } = render(<ScrollButtons {...props} />)

    expect(container.firstChild).toMatchSnapshot
  })

  it('should click on down button', () => {
    render(<ScrollButtons {...props} />)

    userEvent.click(screen.getByRole('button', { name: /down/i }))

    expect(handleDown).toHaveBeenCalledTimes(1)
  })

  it('should click on top button', () => {
    render(<ScrollButtons {...props} />)

    userEvent.click(screen.getByRole('button', { name: /up/i }))

    expect(handleUp).toHaveBeenCalledTimes(1)
  })
})
