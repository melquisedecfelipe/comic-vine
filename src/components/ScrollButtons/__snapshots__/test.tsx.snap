// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScrollButtons /> should render the scroll buttons 1`] = `
.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: Barlow,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-weight: 600;
  font-size: 16px;
  padding: 0 16px;
  cursor: pointer;
  border: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 0.3s ease-in-out;
  transition: 0.3s ease-in-out;
  min-height: 40px;
  max-height: 40px;
  color: #000000;
  background: #FFFFFF;
  border: 1px solid #000000;
}

.c1:focus {
  opacity: 0.75;
}

.c1:hover {
  color: #FFFFFF;
  background: #000000;
}

.c0 {
  position: fixed;
  bottom: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 16px;
  padding: 24px;
}

<div
  class="c0"
>
  <button
    aria-label="Down"
    class="c1"
  >
    Down
  </button>
  <button
    aria-label="Up"
    class="c1"
  >
    Up
  </button>
</div>
`;
